adr/ADR-43.md

```diff
@@ -0,0 +1,87 @@
+# JetStream Per-Message TTL
+
+| Metadata | Value                     |
+|----------|---------------------------|
+| Date     | 2024-07-11                |
+| Author   | @ripienaar                |
+| Status   | Approved                  |
+| Tags     | jetstream, client, server |
+
+## Context and motivation
+
+Streams support a one-size-fits-all approach to message TTL based on the MaxAge setting. This causes any message in the 
+Stream to expire at that age.
+
+There are numerous uses for a per-message version of this limit, some listed below:
+
+ * KV tombstones are a problem in that they forever clog up the buckets with noise, these could have a TTL to make them expire once not useful anymore
+ * Server-applied limits can result in tombstones with a short per message TTL so that consumers can be notified of limits being processed. Useful in KV watch scenarios being notified about TTL removals
+ * A stream may have a general MaxAge but some messages may have infinite retention, think a schema or type hints in a KV bucket that is forever while general keys have TTLs
+ * TODO: user use cases
+
+Related issues [#3268](https://github.com/nats-io/nats-server/issues/3268)
+
+## Per-Message TTL
+
+We will allow a message to supply a TTL as a duration in a header called `Nats-TTL` followed by the duration as 
+nanoseconds, the sever may not support nanosecond level precision and clients can choose to expose TTL as
+seconds or nanoseconds.
+
+The duration will be used by the server to calculate the deadline for removing the message based on its Stream 
+timestamp and the stated duration.
+
+The TTL may not exceed the Stream MaxAge to give administrators a hard cap for the purposes of managing resources.
```

Comment:
@derekcollison I do not know how you'll implement this but I do want to cap both the minimum TTL and Maximum TTL here to avoid issues of 10 year TTLs might cause.

So once you are a bith further down the implementation detail please let me know what to put here.  I reckon smallest minimum might be seconds and the largest days but open to feedback here based on what you are thinking implementation wise.

Reply:
Impl will be on second resolution. You can give us anything but that is granularity we will fire.

For maximum, not sure we need one.

Reply:
I'll stick with being consisent then and go for nanos - clients can present it however they like.

You're 100% sure we dont need to be cautious here that if someone makes a 100 million message stream all with 10 year TTL that this would not cause issues? For sure I'll cap it at MaxAge or forever but just want to be 100% sure we're not creating another point of abuse.

Reply:
The abuse will not be in TTL IMO, also we will allow `does not expire` so. The abuse will come in having millions of these in a stream IMO.

Reply:
OK, made the header second and kept the config option as a duration for consistency.  Minimum for TTL is 1 and uncapped max.